### PR TITLE
Simplify a broader range of java primitive types

### PR DESCRIPTION
**Pull request by reuk**
This is an extension to #1438, which removes circular casts through `byte`, `char`, and `short` in addition to `boolean`. This change has been made because, in working on diffblue/test-gen#1070, I discovered that primitive types shorter than int can create unnecessary circular type-casts in loaded `codet`s.

Suggest for @pkesseli and @thk123 to review.

### CODE REVIEW

**Comment by tautschnig** [comment]
Wouldn't it be possible to avoid generating those typecast expressions in the first place?

**Comment by reuk** [comment]
I don't think so. The issue is that the JVM instruction set treats items shorter than `int` as ints, and relies on the compiler typechecking to get the types to line up. From CBMC's perspective we see something like
- call some function with a `char` parameter
- load that parameter using an iload or similar - now the type appears to be `int`
- from the current function, call another function which takes a `char` parameter, meaning we have to cast our `int` to `char`

File: src/util/simplify_expr.cpp [hunk 1]
```diff
@@ -236,22 +236,34 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
...
-     op_type.id()!=ID_bool)
+  // circular casts through types shorter than `int`
+  const auto check_circular_cast=[&](const typet &type)
   {
-    // casts from boolean to a signed int and back:
-    // (boolean)(int)boolean -> boolean
-    if(expr.op0().id()==ID_typecast && op_type.id()==ID_signedbv)
+    if(expr_type==type &&
+       expr.op0().id()==ID_typecast &&
+       op_type==signedbv_typet(32))
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: I'd suggest to factor out the last two checks as they can be done just once instead of doing them 4 times.

File: src/util/simplify_expr.cpp [hunk 2]
```diff
@@ -236,22 +236,34 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
     return false;
   }
 
-  // eliminate casts to _Bool
-  if(expr_type.id()==ID_c_bool &&
-     op_type.id()!=ID_bool)
+  // circular casts through types shorter than `int`
+  const auto check_circular_cast=[&](const typet &type)
```
Comment (thk123) [inline src/util/simplify_expr.cpp]: Given the length and non-triviality of this lambda I'd suggest pulling it into its own function.

File: src/util/simplify_expr.cpp [hunk 3]
```diff
@@ -236,22 +236,34 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
...
   {
-    // casts from boolean to a signed int and back:
-    // (boolean)(int)boolean -> boolean
-    if(expr.op0().id()==ID_typecast && op_type.id()==ID_signedbv)
+    if(expr_type==type)
     {
-      const auto &typecast=to_typecast_expr(expr.op0());
-      if(typecast.op().type().id()==ID_c_bool)
+      const auto &typecast=expr_dynamic_cast<typecast_exprt>(expr.op0());
```
Comment (thk123) [inline src/util/simplify_expr.cpp]: I come you know for a fact that the first operand is also a type cast?

File: src/util/simplify_expr.cpp [hunk 4]
```diff
@@ -236,22 +236,34 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
...
         expr=typecast.op0();
         return false;
       }
     }
+    return true;
+  };
+  if(op_type==signedbv_typet(32) &&
+     expr.op0().id()==ID_typecast &&
+     !(check_circular_cast(c_bool_typet(8)) &&
+       check_circular_cast(signedbv_typet(8)) &&
```
Comment (thk123) [inline src/util/simplify_expr.cpp]: Why not `unsignedbv_typet(8)`?

File: unit/util/simplify_expr.cpp [hunk 5]
```diff
@@ -61,7 +61,10 @@ TEST_CASE("Simplify const pointer offset")
   REQUIRE(offset_value==1234);
 }
 
-TEST_CASE("Simplify Java boolean -> int -> boolean casts")
+namespace
+{
+
+void test_unnecessary_cast(std::function<typet()> get_type_func)
```
Comment (thk123) [inline unit/util/simplify_expr.cpp]: Why not just take a `const typet &` as the input here?

File: src/util/simplify_expr.cpp [hunk 6]
```diff
@@ -236,22 +236,34 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
...
+      const auto &typecast=expr_dynamic_cast<typecast_exprt>(expr.op0());
+      if(typecast.op().type()==type)
       {
         expr=typecast.op0();
         return false;
       }
     }
+    return true;
+  };
+  if(op_type==signedbv_typet(32) &&
```
Comment (thk123) [inline src/util/simplify_expr.cpp]: Consider breaking this if up as it is hard to tell what the relation and reason of these conditions are

File: src/util/simplify_expr.cpp (hunk 3 above)
Comment (thk123) [reply src/util/simplify_expr.cpp]: I see this in guaranteed by the check in the if, which isn't obvious. I suggest making a method that takes the `typecast_exprt` as an input and returns the modified `exprt`?

File: unit/util/simplify_expr.cpp [hunk 7]
```diff
@@ -70,13 +73,35 @@ TEST_CASE("Simplify Java boolean -> int -> boolean casts")
...
   REQUIRE(simplified.id()==ID_symbol);
-  REQUIRE(simplified.type()==java_boolean_type());
+  REQUIRE(simplified.type()==get_type_func());
   const auto &symbol=to_symbol_expr(simplified);
   REQUIRE(symbol.get_identifier()=="foo");
 }
+
+} // namespace
+
+TEST_CASE("Simplify Java boolean -> int -> boolean casts")
```
Comment (thk123) [inline unit/util/simplify_expr.cpp]: You should add tests for the cases that shouldn't be simplified ideally (e.g. a typecast from an int to a short)

**Review by thk123 — CHANGES_REQUESTED** [review]
New code is kind of hard to follow, I've made some suggestions but feel free to come up with a better way.

File: src/util/simplify_expr.cpp (hunk 4 above)
Comment (reuk) [reply src/util/simplify_expr.cpp]: These are supposed to correspond to the shorter-than-int types in java_types.h. However, `util` shouldn't depend on other modules so I don't know if it's OK to include that header here.

File: unit/util/simplify_expr.cpp (hunk 5 above)
Comment (reuk) [reply unit/util/simplify_expr.cpp]: Good point

File: src/util/simplify_expr.cpp (hunk 2 above)
Comment (reuk) [reply src/util/simplify_expr.cpp]: Turns out it can be written as a for-loop, which means we don't need to pass a bunch of local variables into a whole new function.

File: src/util/simplify_expr.cpp (hunk 3 above)
Comment (reuk) [reply src/util/simplify_expr.cpp]: Added a comment explaining how we know the `expr_dynamic_cast` is non-throwing

**Comment by reuk** [comment]
I've responded to feedback, please re-review

File: src/util/simplify_expr.cpp [hunk 8]
```diff
@@ -236,22 +236,36 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
...
+      c_bool_typet(8),
+      signedbv_typet(8),
+      signedbv_typet(16),
+      unsignedbv_typet(16),
+    };
+    for(const auto &type : types_to_check)
     {
-      const auto &typecast=to_typecast_expr(expr.op0());
-      if(typecast.op().type().id()==ID_c_bool)
+      if(expr_type==type)
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: No need for a `for` loop:
```
if(expr_type==c_bool_typet(8) ||
   expr_type==signedbv_typet(8) ||
   expr_type==signedbv_typet(16) ||
   expr_type==unsignedbv_typet(16))
```
and then just replace `type` by `expr_type` below.

File: src/util/simplify_expr.cpp [hunk 9]
```diff
@@ -236,22 +236,29 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
...
+     expr.op0().id()==ID_typecast)
+  {
+    if(expr_type==c_bool_typet(8) ||
+       expr_type==signedbv_typet(8) ||
+       expr_type==signedbv_typet(16) ||
+       expr_type==unsignedbv_typet(16))
     {
-      const auto &typecast=to_typecast_expr(expr.op0());
-      if(typecast.op().type().id()==ID_c_bool)
+      // We checked that the id ws ID_typecast in the enclosing `if`
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: Nit picking: typo "ws" -> "was"

**Review by tautschnig — APPROVED** [review]
Looks good to me, but indeed @pkesseli should sign off on this as well.

File: src/util/simplify_expr.cpp (hunk 4 above)
Comment (thk123) [reply src/util/simplify_expr.cpp]: I see the logic (and I don't think you should include the header0, but would there be anything wrong from a C perspective of disregarding a cast from an unsigned 16 -> unsigned 8 -> unsigned 16 for example? Feel free to say out of scope :)

File: src/util/simplify_expr.cpp (hunk 2 above)
Comment (thk123) [reply src/util/simplify_expr.cpp]: Where's the for loop?

File: unit/util/simplify_expr.cpp [hunk 10]
```diff
@@ -61,22 +61,67 @@ TEST_CASE("Simplify const pointer offset")
...
+
+    {
+      const exprt simplified=simplify_expr(
+        typecast_exprt(symbol_exprt("foo", java_int_type()), type),
+        namespacet(symbol_tablet()));
+
+      REQUIRE(simplified.id()==ID_typecast);
+      REQUIRE(simplified.type()==type);
+    }
+  }
```
Comment (thk123) [inline unit/util/simplify_expr.cpp:104]: You might consider a test that involves much more nesting `(int)(short)(int)(short)` for example

**Review by thk123 — APPROVED** [review]
Seems fine though I'm a bit concerned you mentioned converting code into a loop and there is no loop...

File: unit/util/simplify_expr.cpp (hunk 10 above)
Comment (reuk) [reply unit/util/simplify_expr.cpp:104]: Done.

**Comment by reuk** [comment]
> I'm a bit concerned you mentioned converting code into a loop and there is no loop...

@tautschnig pointed out that the loop isn't necessary after all, so I removed it

File: src/util/simplify_expr.cpp [hunk 11]
```diff
@@ -236,22 +236,29 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
...
+  {
+    if(expr_type==c_bool_typet(8) ||
+       expr_type==signedbv_typet(8) ||
+       expr_type==signedbv_typet(16) ||
+       expr_type==unsignedbv_typet(16))
     {
-      const auto &typecast=to_typecast_expr(expr.op0());
-      if(typecast.op().type().id()==ID_c_bool)
+      // We checked that the id was ID_typecast in the enclosing `if`
+      const auto &typecast=expr_dynamic_cast<typecast_exprt>(expr.op0());
```
Comment (pkesseli) [inline src/util/simplify_expr.cpp]: Since you already checked, this could be a static cast: `to_typecast_expr(...)`

**Review by pkesseli — APPROVED** [review]
Looks good, small static cast recommendation.

**Comment by reuk** [comment]
@thk123 I've responded to all feedback, could you merge this please
